### ISSUE TITLE
Need to have contextual property groups (relative to Git branches)

### ISSUE THREAD

**Issue by FALLAI-Denis** [IBM/zopeneditor-about#97]
## Description of the enhancement requested

Extract from the documentation:

```
Z Open Editor searches for property groups and user build settings in the following order and uses only the first set it finds:
1. zapp.yaml in the workspace directory.
2. zapp.json in the workspace directory.
3. .vscode/settings.json in the workspace directory, which is the VS Code Workspace Setting you enter via the Preferences menu.
4. Your user settings that you enter via the Preferences menu, which are stored in a different user-specific location depending on your operating system.
```

This means that there can only be one and only one context of use of property groups.
In addition, the zapp.yaml / zapp.json files being in the Workspace, and therefore in the Git repository, they are common to all developers.

We use the zapp.yaml solution because it is independent of the VS Code configuration.

However, we need to manage multiple contexts of property groups of type "mvs", because we can develop and compile to multiple contexts:
* New version in development
* Maintenance of the production version

And therefore we need to indicate different COPYBOOKs (for COBOL) searches on "mvs" depending on the context.

The "local" COPYBOOKs are necessarily the right ones because they are managed in the branch of the Git Repository, and therefore in the right version.

Note: we are using the Source Management Endevor, and multiple Entry Stage and Mappings are defined, with different SYSLIB references.

For us the "mvs" property groups context depends on the kind of the Git branch in which we are developing:
* "hotfix/xxx" branch: maintenance of the production version
* "feature/xxx" branch, created from "master" (or "develop"): next version in development
* "feature/xxx" branch, created from "release/xxx": future version in developement (many "release/xxx" branches can exist)

When developing (and compiling) for the production version, there is no reason of taking into consideration the COPYBOOKs that are planned for a future version under development, and vice versa.

Ideally, for editing the source code and searching for "mvs" COPYBOOKs, we should be able to select one (or more) property groups according to the Git branch on which we are positioned (subject to using Git, which is our case).

An alternative would be to be able to define property groups at User level, which would overload (and not replace), and would be taken into account as a priority over property groups declared at Workspace level, (which is not the current behavior). .

**Comment by phaumer** [comment]
This is great feedback. The ZAPP files was created as a file stored in your workspace so that it can be managed in Git as another source file and therefore that it can be branched as well. So if you have a different versions of the ZAPP file in each of your branches that defines different MVS property groups, would that not be a valid solution?

**Comment by FALLAI-Denis** [comment]
The zapp file being tracked by Git, and however (for our operations) dependent and different for each branch, the developer will have to intervene on its content after each merge from another branch, with the risk of omissions and the consequence of targeting a bad build context.

**Comment by phaumer** [comment]
But isn't that true for all source files that require merging? Our goal is to provide an equivalent to package.json, pom.xml, build.gradle etc. with a file that can be treated as part of the source code in git and version controlled just like any other source file empowering the developer to do devops and ultimately control how the application is build and deployed (for testing).

**Comment by FALLAI-Denis** [comment]
Difference between "zapp.yaml" and other files :

- "zapp.yaml" contains settings common to all developers and used in a specific context (the branch context). It do'nt must be merged from branch to branch (use of .gitignore ?)

- other files are artefacts of the application. They must be merged from branch to branch. 

Putting "zapp.yaml" in .gitignore can be a solution, but in thèse case each user must manually update content of "zapp.yaml" to reflect ils branch context. 

So we would like to be able to define properties groups filtered by branch name patterns (regexp) :
- for "master" (unique) branch
- for "develop" (unique) branch
- for "hotfix/*" (all) branches 
- for "release/xyz" (specific) branch 
- for "release/abc" (specific) branch
- for "feature/uvw" (specific) branch
- for "feature/*" (all other) branches

Notice the specific filtering for feature branches : a default context and some specific contexts. 

This is similar to parallel development.

**Comment by phaumer** [comment]
Thanks @FALLAI-Denis, but the vision of ZAPP was that it is managed in the SCM with the source and just like the source be agnostic of the specific SCM technology being used. Just like a gradle or package.json file. When you make changes in one branch and merge that into another the intent of a "merge" is to move all changes from one branch to the other. If more changes are required on top of that, or changes to be rejected, or conflicts to be resolved then those should be done as part of the merge operation. 

For your use case I could envision end-user scripts that wrap git operations into larger macros that retrieve and copy gitignored files into place  based on a central specification file that encodes the info in your bullet point list perhaps?

**Comment by FALLAI-Denis** [comment]
Thanks @phaumer.

Our main problem concerns the search for COPYBOOKs / INCLUDEs on the host system (mvs) in the case of developing several versions in parallel:
- a production version on which maintenance must be applied
- a version under development, which will become the next version in production
- other versions which will succeed the version under development

Today, in our organization, all versions are developed in the same Git repository, but in different branches:
- deliveries on the production version are made on "hotfix / *" branches
- deliveries to the version under development are made to the "master" branch
- deliveries on future versions are made on "release / *" branches (which will integrate the "master" branch when the time comes)
 
Developments are carried out in "feature/*" (or "hotfix/*") branches which are created from "master" or "release/*" branches.
When editing source code in IBM Z Open Editor, some COPYBOOKs / INCLUDEs are looked for in the Git repository (local), in the current branch, because they are part of the application, but other COPYBOOKs / INCLUDEs are searched on the host system (mvs), because they are not part of the application, (interface to other applications).

Depending on the version considered, (maintenance of the production version, development of the next version, development of a future version), the search for COPYBOOKs / INCLUDEs on the host system (mvs) is not done on the same targets, (syslib list).

The current implementation of the zapp.yaml file does not allow us to specify searches that depend on the affected version, so for us based on the current Git branch.

One solution would be to have mono version Git repositories, and therefore to have as many Git repositories as there are versions, which would then complicate the postponements between versions, (postponements between Git repositories), especially in a "continuous integration" mode. / continuous deployment ".

So, we don't know how to deal with the need to develop multiple versions in parallel with the zapp.yaml file wich define search for COPYBOOKs / INCLUDEs.
